r2dbc/src/main/java/com/google/cloud/sql/r2dbc/GcpConnectionFactoryProvider.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2020 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.r2dbc;
+
+import java.util.Properties;
+import java.util.function.Function;
+
+import com.google.cloud.sql.core.CoreSocketFactory;
+import com.google.cloud.sql.core.SslData;
+import dev.miku.r2dbc.mysql.MySqlConnectionFactoryProvider;
+import dev.miku.r2dbc.mysql.constant.SslMode;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.r2dbc.postgresql.PostgresqlConnectionFactoryProvider;
+import io.r2dbc.postgresql.client.SSLMode;
+import io.r2dbc.spi.ConnectionFactory;
+import io.r2dbc.spi.ConnectionFactoryOptions;
+import io.r2dbc.spi.ConnectionFactoryProvider;
+import io.r2dbc.spi.Option;
+
+import static io.r2dbc.spi.ConnectionFactoryOptions.*;
+
+/**
+ * {@link ConnectionFactoryProvider} for proxied access to GCP Postgres and MySQL instances.
+ */
+public class GcpConnectionFactoryProvider implements ConnectionFactoryProvider {
+
+	private static final boolean MYSQL_AVAILABLE = isClassPresent("dev.miku.r2dbc.mysql.MySqlConnectionFactoryProvider");
+
+	private static final boolean POSTGRES_AVAILABLE = isClassPresent("io.r2dbc.postgresql.PostgresqlConnectionFactoryProvider");
+
+	/**
+	 * MySQL driver option value.
+	 */
+	public static final String MYSQL_DRIVER = "mysql";
+
+	/**
+	 * Postgres driver option value.
+	 */
+	public static final String POSTGRESQL_DRIVER = "postgresql";
+
+	/**
+	 * Legacy postgres driver option value.
+	 */
+	public static final String LEGACY_POSTGRESQL_DRIVER = "postgres";
+
+	/**
+	 * Option to configure the {@code proxyPort} if {@code port} is not configured.
+	 */
+	public static final Option<Object> PROXY_PORT = Option.valueOf("proxyPort");
+
+	@Override
+	public ConnectionFactory create(ConnectionFactoryOptions connectionFactoryOptions) {
```

Comment:
Given we need to split into two artifacts, I think it makes sense to move this logic into CoreSocketFactory and just have thin wrappers in each artifact.